sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/thetasketchesAggregates.scala

```diff
@@ -0,0 +1,652 @@
+  since = "4.1.0")
+// scalastyle:on line.size.limit
+case class ThetaSketchAgg(
+    left: Expression,
+    right: Expression,
+    mutableAggBufferOffset: Int = 0,
+    inputAggBufferOffset: Int = 0)
+    extends TypedImperativeAggregate[ThetaSketchState]
+    with BinaryLike[Expression]
+    with ExpectsInputTypes {
+
+  // ThetaSketch config - mark as lazy so that they're not evaluated during tree transformation.
+
+  lazy val lgNomEntries: Int = {
+    val lgNomEntriesInput = right.eval().asInstanceOf[Int]
+    ThetaSketchUtils.checkLgNomLongs(lgNomEntriesInput)
+    lgNomEntriesInput
+  }
+
+  // Constructors
+
+  def this(child: Expression) = {
+    this(child, Literal(ThetaSketchUtils.DEFAULT_LG_NOM_LONGS), 0, 0)
+  }
+
+  def this(child: Expression, lgNomEntries: Expression) = {
+    this(child, lgNomEntries, 0, 0)
+  }
+
+  def this(child: Expression, lgNomEntries: Int) = {
+    this(child, Literal(lgNomEntries), 0, 0)
+  }
+
+  // Copy constructors required by ImperativeAggregate
+
+  override def withNewMutableAggBufferOffset(newMutableAggBufferOffset: Int): ThetaSketchAgg =
+    copy(mutableAggBufferOffset = newMutableAggBufferOffset)
+
+  override def withNewInputAggBufferOffset(newInputAggBufferOffset: Int): ThetaSketchAgg =
+    copy(inputAggBufferOffset = newInputAggBufferOffset)
+
+  override protected def withNewChildrenInternal(
+      newLeft: Expression,
+      newRight: Expression): ThetaSketchAgg =
+    copy(left = newLeft, right = newRight)
+
+  // Overrides for TypedImperativeAggregate
+
+  override def prettyName: String = "theta_sketch_agg"
+
+  override def inputTypes: Seq[AbstractDataType] =
+    Seq(
+      TypeCollection(
+        IntegerType,
+        LongType,
+        FloatType,
+        DoubleType,
+        StringTypeWithCollation(supportsTrimCollation = true),
+        BinaryType,
+        ArrayType(IntegerType),
+        ArrayType(LongType)),
+      IntegerType)
+
+  override def dataType: DataType = BinaryType
+
+  override def nullable: Boolean = false
+
+  /**
+   * Instantiate an UpdateSketch instance using the lgNomEntries param.
+   *
+   * @return
+   *   an UpdateSketch instance wrapped with UpdatableSketchBuffer
+   */
+  override def createAggregationBuffer(): ThetaSketchState = {
+    val builder = new UpdateSketchBuilder
+    builder.setLogNominalEntries(lgNomEntries)
+    UpdatableSketchBuffer(builder.build)
+  }
+
+  /**
+   * Evaluate the input row and update the UpdateSketch instance with the row's value. The update
+   * function only supports a subset of Spark SQL types, and an exception will be thrown for
+   * unsupported types.
+   *
+   * @param updateBuffer
+   *   A previously initialized UpdateSketch instance
+   * @param input
+   *   An input row
+   */
+  override def update(updateBuffer: ThetaSketchState, input: InternalRow): ThetaSketchState =
+    updateBuffer match {
+      case UpdatableSketchBuffer(sketch) =>
+        val v = left.eval(input)
+        if (v != null) {
```

Comment:
you can simplify the code by just removing this and adding a `case _ if v == null =>`, removing an indentation level.

Reply:
yep makes sense, will fix those